core/src/main/scala/akka/kafka/internal/ConsumerStage.scala

```diff
@@ -341,15 +341,18 @@ private[kafka] trait PromiseControl extends GraphStageLogic with Control {
 
 private[kafka] trait MetricsControl extends Control {
 
-  protected def consumer: ActorRef
+  protected def executionContext: ExecutionContext
+  protected def consumerFuture: Future[ActorRef]
 
   def metrics: Future[Map[MetricName, Metric]] = {
     import akka.pattern.ask
-
     import scala.concurrent.duration._
-    consumer
-      .?(RequestMetrics)(Timeout(1.minute))
-      .mapTo[ConsumerMetrics]
-      .map(_.metrics)(ExecutionContexts.sameThreadExecutionContext)
+    consumerFuture
+      .flatMap { consumer =>
+        consumer
+          .ask(RequestMetrics)(Timeout(1.minute))
+          .mapTo[ConsumerMetrics]
+          .map(_.metrics)(ExecutionContexts.sameThreadExecutionContext)
+      }(executionContext)
```

Comment:
@ennru I'm calling `metrics` after materializing a stream and I'm getting an exception like this:
```
java.lang.IllegalStateException: not yet initialized: only setHandler is allowed in GraphStageLogic constructor. To access materializer use Source/Flow/Sink.setup factory
	at akka.stream.stage.GraphStageLogic.interpreter(GraphStage.scala:393)
	at akka.stream.stage.GraphStageLogic.materializer(GraphStage.scala:403)
	at akka.kafka.internal.BaseSingleSourceLogic.executionContext(BaseSingleSourceLogic.scala:34)
	at akka.kafka.internal.MetricsControl.metrics(ControlImplementations.scala:88)
	at akka.kafka.internal.MetricsControl.metrics$(ControlImplementations.scala:78)
	at akka.kafka.internal.BaseSingleSourceLogic.metrics(BaseSingleSourceLogic.scala:26)
```
The flow takes messages from a `KafkaConsumerActor` and sends them to another actor, the whole thing looks like this:
```
      val sink: Sink[Output, NotUsed] =
        ActorSink.actorRefWithAck[Output, SinkActorCommandType, Ack](...)

      val control: Consumer.Control =
        Consumer.plainExternalSource[String, Array[Byte]](kafkaConsumerActor, subscription)
          .map { consumerRecord =>
            processRecord(consumerRecord)
          }
          .to(sink)
          .run()

      control.metrics
```
Did the race condition just move to a different place now?

Reply:
OK, looks like somebody already raised the issue, #847